core/src/main/java/com/datastrato/gravitino/storage/relational/service/UserMetaService.java

```diff
@@ -221,6 +224,39 @@ public <E extends Entity & HasIdentifier> UserEntity updateUser(
     return newEntity;
   }
 
+  public List<UserEntity> listUsersByNamespace(Namespace namespace) {
+    AuthorizationUtils.checkUserNamespace(namespace);
+
+    List<UserEntity> userEntities = Lists.newArrayList();
+    String metalakeName = namespace.level(0);
+    List<UserPO> userPOs = Lists.newArrayList();
+    AtomicLong metalakeId = new AtomicLong();
+    SessionUtils.doMultipleWithoutCommit(
+        () -> {
+          Long id =
+              SessionUtils.doWithoutCommitAndFetchResult(
+                  MetalakeMetaMapper.class,
+                  mapper -> mapper.selectMetalakeIdMetaByName(metalakeName));
+          if (id == null) {
+            throw new NoSuchEntityException(
+                NoSuchEntityException.NO_SUCH_ENTITY_MESSAGE,
+                Entity.EntityType.METALAKE.name().toLowerCase(),
+                metalakeName);
+          }
+          metalakeId.set(id);
+        },
+        () ->
+            userPOs.addAll(
+                SessionUtils.doWithoutCommitAndFetchResult(
+                    UserMetaMapper.class,
+                    mapper -> mapper.listUserPOsByMetalakeId(metalakeId.get()))));
+    for (UserPO userPO : userPOs) {
+      List<RolePO> rolePOs = RoleMetaService.getInstance().listRolesByUserId(userPO.getUserId());
+      userEntities.add(POConverters.fromUserPO(userPO, rolePOs, namespace));
+    }
```

Comment:
the num of users > 1w is common situation, this may execute for a long time, maybe we should consider using cache in `listUsers`

Reply:
cc @yuqi1129 Maybe we can have a common cache layer for this.

Reply:
I think it is better to use a paging query.